crates/wallet/src/wallet/mod.rs

```diff
@@ -366,231 +383,157 @@ impl Wallet {
-        let mut index = KeychainTxOutIndex::<KeychainKind>::default();
+        let chain = LocalChain::from_changeset(changeset.local_chain)
+            .map_err(|_| LoadError::MissingGenesis)?;
+
+        let mut descriptor_keymap = params.descriptor_keymap;
         let descriptor = changeset
-            .indexed_tx_graph
-            .indexer
-            .keychains_added
-            .get(&KeychainKind::External)
-            .ok_or(LoadError::MissingDescriptor(KeychainKind::External))?
-            .clone();
+            .descriptor
+            .ok_or(LoadError::MissingDescriptor(KeychainKind::External))?;
+        check_wallet_descriptor(&descriptor).map_err(LoadError::Descriptor)?;
+
+        let mut change_descriptor_keymap = params.change_descriptor_keymap;
         let change_descriptor = changeset
-            .indexed_tx_graph
-            .indexer
-            .keychains_added
-            .get(&KeychainKind::Internal)
-            .ok_or(LoadError::MissingDescriptor(KeychainKind::Internal))?
-            .clone();
+            .change_descriptor
+            .ok_or(LoadError::MissingDescriptor(KeychainKind::Internal))?;
+        check_wallet_descriptor(&change_descriptor).map_err(LoadError::Descriptor)?;
+
+        // checks
+        if let Some(exp_network) = params.check_network {
+            if network != exp_network {
+                return Err(LoadError::Mismatch(LoadMismatch::Network {
+                    loaded: network,
+                    expected: exp_network,
+                }));
+            }
+        }
+        if let Some(exp_genesis_hash) = params.check_genesis_hash {
+            if chain.genesis_hash() != exp_genesis_hash {
+                return Err(LoadError::Mismatch(LoadMismatch::Genesis {
+                    loaded: chain.genesis_hash(),
+                    expected: exp_genesis_hash,
+                }));
+            }
+        }
+        if let Some(exp_descriptor) = params.check_descriptor {
+            let (exp_descriptor, keymap) =
+                (exp_descriptor)(&secp, network).map_err(LoadError::Descriptor)?;
+            descriptor_keymap.extend(keymap);
+
+            if descriptor.descriptor_id() != exp_descriptor.descriptor_id() {
+                return Err(LoadError::Mismatch(LoadMismatch::Descriptor {
+                    keychain: KeychainKind::External,
+                    loaded: descriptor,
+                    expected: exp_descriptor,
+                }));
+            }
+        }
+        if let Some(exp_change_descriptor) = params.check_change_descriptor {
+            let (exp_change_descriptor, keymap) =
+                (exp_change_descriptor)(&secp, network).map_err(LoadError::Descriptor)?;
+            change_descriptor_keymap.extend(keymap);
+
+            if change_descriptor.descriptor_id() != exp_change_descriptor.descriptor_id() {
+                return Err(LoadError::Mismatch(LoadMismatch::Descriptor {
+                    keychain: KeychainKind::External,
```

Comment:
```suggestion
                    keychain: KeychainKind::Internal,
```

Reply:
Also this should be been caught by a unit test.

Reply:
I pushed a commit to fix this: 8875c92ec114ba9ca0ea8402d65d7a236566b618